### PR TITLE
SMC Improvements

### PR DESCRIPTION
**Pull request by CoryMcCartan**
This PR

1. Adds a missing correction term to the SMC algorithm
2. Adds parallel capacity (within and across sampling runs) to SMC
3. Provides new diagnostics functions and capabilities, with friendly, human-readable output
4. Performs general cleanup, etc., towards `redist` 4.0

### CODE REVIEW

File: R/redist_shortburst.R [hunk 1]
```diff
@@ -129,7 +129,7 @@ redist_shortburst = function(map, score_fn=NULL, stop_at=NULL,
     if (is.null(init_plan)) init_plan = as.integer(as.factor(get_existing(map)))
     if (length(init_plan) == 0L || isTRUE(init_plan == "sample")) {
         init_plan = as.integer(get_plans_matrix(
-            redist_smc(map, 10, counties, resample=FALSE, ref_name=FALSE, silent=TRUE))[, 1])
+            redist_smc(map, 10, counties, resample=FALSE, ref_name=FALSE, silent=TRUE, cores=1))[, 1])
```
Comment (christopherkenny) [inline R/redist_shortburst.R:132]: ncores vs cores

File: R/redist_smc.R [hunk 2]
```diff
@@ -104,21 +116,24 @@
...
 #'
 #' @concept simulate
 #' @md
 #' @order 1
 #' @export
 redist_smc = function(map, nsims, counties=NULL, compactness=1, constraints=list(),
-                      resample=TRUE, constraint_fn=function(m) rep(0, ncol(m)),
-                      init_particles=NULL, n_steps=NULL,
-                      adapt_k_thresh=0.985, seq_alpha=0.2+0.3*compactness,
+                      resample=TRUE, runs=1L, cores=0L, init_particles=NULL,
```
Comment (christopherkenny) [inline R/redist_smc.R:132]: Can we avoid adding a new term like runs?

File: R/redist_smc.R [hunk 3]
```diff
@@ -278,6 +380,79 @@ redist_smc = function(map, nsims, counties=NULL, compactness=1, constraints=list
...
+#' iowa_map = redist_map(iowa, existing_plan=cd_2010, pop_tol=0.05)
+#' plans = redist_mergesplit_parallel(iowa_map, nsims=200, chains=2, silent=TRUE) %>%
+#'     mutate(dem = group_frac(iowa_map, dem_08, dem_08 + rep_08)) %>%
+#'     number_by(dem)
+#' redist_smc_ci(plans, dem)
+#'
+#' @md
+#' @concept analyze
+#' @export
+redist_smc_ci = function(plans, x, district=1L, conf=0.9) {
```
Comment (christopherkenny) [inline R/redist_smc.R:423]: Is default 0.9 best?

File: R/tidy.R [hunk 4]
```diff
@@ -644,20 +635,21 @@ prec_assignment = function(prec, .data=cur_plans()) {
...
 #' @param sampled_only if `TRUE`, do not include reference plans.
+#' @param cores the number of parallel cores to use in the computation.
 #'
 #' @return a symmetric matrix the size of the number of precincts.
 #'
 #' @concept analyze
 #' @md
 #' @export
-prec_cooccurrence = function(plans, which=NULL, sampled_only=TRUE) {
+prec_cooccurrence = function(plans, which=NULL, sampled_only=TRUE, cores=1) {
```
Comment (christopherkenny) [inline R/tidy.R:645]: ncores/cores

File: explore/basic_threads.R [hunk 5]
```diff
@@ -0,0 +1,15 @@
+library(Rcpp)
```
Comment (christopherkenny) [inline explore/basic_threads.R:1]: drop file?

File: explore/devs.R [hunk 6]
```diff
@@ -0,0 +1,20 @@
+library(tidyverse)
```
Comment (christopherkenny) [inline explore/devs.R:1]: drop file?

File: explore/labeling.R [hunk 7]
```diff
@@ -0,0 +1,40 @@
+library(ggplot2)
```
Comment (christopherkenny) [inline explore/labeling.R:1]: drop file?

File: explore/stderr.R [hunk 8]
```diff
@@ -0,0 +1,147 @@
+library(dplyr)
```
Comment (christopherkenny) [inline explore/stderr.R:1]: drop file?

File: explore/threads.R [hunk 9]
```diff
@@ -0,0 +1,17 @@
+devtools::load_all(here::here("."))
```
Comment (christopherkenny) [inline explore/threads.R:1]: drop file?

File: src/labeling.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            max_x = xchild[i];
+        }
+    }
+
+    double accuml = 0.0;
+    for (int i = 0; i < n; i++) {
+        accuml += std::exp(xchild[i] - max_x);
+    }
+
+    // TODO subtract lgamma(n+1)
```
Comment (christopherkenny) [inline src/labeling.cpp:26]: remaining TODO

File: tests/testthat/test_smc.R [hunk 11]
```diff
@@ -80,13 +80,13 @@ test_that("Partial sampling works accurately", {
     ref_plans = plans_10[, redist.parity(plans_10, pop) <= 0.01]
     log_st_ref = round(log_st_map(adj, ref_plans, rep(1L, 25), 3L), 5)
 
-    out1 = redist_smc(set_pop_tol(fl_map, 0.01), 2000, compactness=0,
-                      n_steps=1, adapt_k_thresh=0.99, seq_alpha=0.3,
+    out1 = redist_smc(set_pop_tol(fl_map, 0.01), 3000, compactness=0,
+                      n_steps=1, adapt_k_thresh=1, seq_alpha=0.5,
                       resample=T, silent=T) %>%
```
Comment (christopherkenny) [inline tests/testthat/test_smc.R:85]: T

**Review by christopherkenny — CHANGES_REQUESTED** [review]
Pretty much only cosmetics

**Comment by kuriwaki** [comment]
Additional question: how does set.seed work with `runs` in SMC. Guidance on how to make SMC with multiple runs replicable with set.seed would be helpful.

File: R/classify_compare.R [hunk 12]
```diff
@@ -64,7 +65,7 @@
 #' @concept analyze
 #' @export
 compare_plans = function(plans, set1, set2, shp=NULL, plot="fill", thresh=0.1,
-                         labs=c("Set 1", "Set 2")) {
+                         labs=c("Set 1", "Set 2"), cores=1) {
```
[the remark this answers is not in the document]
Comment (CoryMcCartan) [reply R/classify_compare.R:68]: oh I guess we are using ncores elsewhere :( 
can change over

File: R/redist_ms.R [hunk 13]
```diff
@@ -189,11 +190,11 @@ redist_mergesplit = function(map, nsims, warmup=max(100, nsims %/% 2), thin=1L,
                      pop_bounds[2], pop_bounds[1], pop_bounds[3], compactness,
                      constraints, adapt_k_thresh, k, thin, verbosity)
 
-    plans <- algout$plans
+    storage.mode(algout$plans) = "integer"
```
[the remark this answers is not in the document]
Comment (CoryMcCartan) [reply R/redist_ms.R:193]: the C++ is supposed to return an integer matrix but turns out it gets converted to a numeric matrix.
this line converts back to an integer, which halves the storage space consumed by the matrix

File: R/redist_smc.R (hunk 2 above)
Comment (CoryMcCartan) [reply R/redist_smc.R:132]: what would you suggest?

File: R/redist_smc.R (hunk 3 above)
Comment (CoryMcCartan) [reply R/redist_smc.R:423]: I think so. 95% aren't very stable. 5th to 95th percentile seems good

File: explore/basic_threads.R (hunk 5 above)
Comment (CoryMcCartan) [reply explore/basic_threads.R:1]: `explore/` is in `.Rbuildignore`. I think it's useful to have some scratch work - type files that are also version controlled. Bit of a testing workbench for new features etc.

File: src/labeling.cpp (hunk 10 above)
Comment (CoryMcCartan) [reply src/labeling.cpp:26]: yeah I need to decide this

File: R/redist_ms.R (hunk 13 above)
[the remark this answers is not in the document]
Comment (christopherkenny) [reply R/redist_ms.R:193]: Makes sense on storage. Very interesting on C++ returns!

File: R/redist_smc.R (hunk 2 above)
Comment (christopherkenny) [reply R/redist_smc.R:132]: Is chains appropriate? If not, then runs makes sense. Just trying to avoid any sort of new glossary situation.

Comment (CoryMcCartan) [reply R/redist_smc.R:132]: yeah in the `plans` object the column is called `chains` but it's definitely not a markov chain so I'm reluctant to adopt it in the actual function.  Dunno if it needs  a glossary entry if it's just for SMC

**Comment by CoryMcCartan** [comment]
> Additional question: how does set.seed work with `runs` in SMC. Guidance on how to make SMC with multiple runs replicable with set.seed would be helpful.

@kuriwaki yes, good point.  Looking into this. Goal would be for a single `set.seed` call to determine all of the runs' output

File: R/redist_smc.R [hunk 14]
```diff
@@ -294,6 +469,7 @@ redist_quantile_trunc = function(x) pmin(x, quantile(x, 1 - length(x)^(-0.5)))
 #' @concept post
 #' @export
 redist.smc_is_ci = function(x, wgt, conf=0.99) {
+    .Deprecated("redist_smc_ci")
```
Comment (christopherkenny) [inline R/redist_smc.R:472]: Can we point to the new fn?

**Review by christopherkenny — APPROVED** [review]
one tiny thing, otherwise looks good!

**Pull request merged**